statsd/aggregator.go

```diff
@@ -3,61 +3,48 @@ package statsd
 import (
 	"fmt"
 	"math"
-	"runtime"
 	"sort"
-	"strings"
-	"sync"
 	"time"
 
-	"github.com/atlassian/gostatsd/backend"
 	"github.com/atlassian/gostatsd/types"
 
 	log "github.com/Sirupsen/logrus"
 )
 
-// metricAggregatorStats is a bookkeeping structure for statistics about a MetricAggregator.
-type metricAggregatorStats struct {
-	BadLines       int64
-	LastMessage    time.Time
-	LastFlush      time.Time
-	LastFlushError time.Time
-	NumStats       int
-	ProcessingTime time.Duration
-}
+// ProcessFunc is a function that gets executed by Aggregator with its state passed into the function.
+type ProcessFunc func(*types.MetricMap)
 
-// MetricAggregator is an object that aggregates statsd metrics.
+// Aggregator is an object that aggregates statsd metrics.
 // The function NewMetricAggregator should be used to create the objects.
 //
-// Incoming metrics should be sent to the MetricQueue channel.
-type MetricAggregator struct {
-	sync.Mutex
-	ExpiryInterval    time.Duration      // How often to expire metrics
-	FlushInterval     time.Duration      // How often to flush metrics to the sender
-	LastFlush         time.Time          // Last time the metrics where aggregated
-	MaxWorkers        int                // Number of workers to metrics queue
-	MetricQueue       chan *types.Metric // Queue on which metrics are received
-	PercentThresholds []float64
-	Senders           []backend.MetricSender // The sender to which metrics are flushed
-	Stats             metricAggregatorStats
-	Tags              types.Tags // Tags to add to all metrics
+// Incoming metrics should be passed via ReceiveMetric function.
+type Aggregator interface {
+	ReceiveMetric(*types.Metric, time.Time)
+	Flush(func() time.Time) *types.MetricMap
+	Process(ProcessFunc)
+	Reset(time.Time)
+}
+
+type metricAggregator struct {
+	expiryInterval    time.Duration // How often to expire metrics
+	lastFlush         time.Time     // Last time the metrics where aggregated
+	percentThresholds []float64
+	systemTags        string // Tags to add to system metrics
 	types.MetricMap
 }
 
-// NewMetricAggregator creates a new MetricAggregator object.
-func NewMetricAggregator(senders []backend.MetricSender, percentThresholds []float64, flushInterval time.Duration, expiryInterval time.Duration, maxWorkers int, tags []string) *MetricAggregator {
-	a := MetricAggregator{}
+// NewMetricAggregator creates a new Aggregator object.
+func NewMetricAggregator(percentThresholds []float64, flushInterval, expiryInterval time.Duration, systemTags []string) Aggregator {
```

Comment:
Yes, that is how we used to do. I changed it a bit, hence the rename. Let me walk you through.

* This is how we create an implementation for `AggregatorFactory` interface and pass it to `Dispatcher`:
```go
	// 1. Start the Dispatcher
	factory := agrFactory{
		percentThresholds: percentThresholds,
		flushInterval:     s.FlushInterval,
		expiryInterval:    s.ExpiryInterval,
		systemTags:        s.DefaultTags,
	}
	dispatcher := NewDispatcher(s.MaxWorkers, s.MaxQueueSize, &factory)
```
As you can see we take `DefaultTags` from configuration and pass them as `systemTags`.
* They are called `systemTags` at this stage because `agrFactory` will enrich them and pass to `NewMetricAggregator()`. These tags are applied to system metrics that we send per **each** aggregator (`num_stats` and `processing_time` per aggragator instance):
```go
func (af *agrFactory) Create() Aggregator {
	tags := make([]string, 0, len(af.systemTags)+1)
	tags = append(tags, af.systemTags...)
	tags = append(tags, fmt.Sprintf("aggregator_%d", af.workerNumber))
	af.workerNumber++
	return NewMetricAggregator(af.percentThresholds, af.flushInterval, af.expiryInterval, tags)
}
```
* So we can see metrics per-aggregator and totals.

I'm not insisting at all, I'll rename it. Just wanted to explain why I decided to rename. Default tags are applied to received metrics by receiver, not by aggregator. Aggregator only applies these tags to its own metrics (`num_stats` and `processing_time`).